### PR TITLE
Move `sql/sqlite` to `backends/sqlite`

### LINKED ISSUE #2420: Move SQLITE backend to the backends/ directory
**Move the `sql/sqlite/` backend to `backends/sqlite`.**

The current Ibis directory structure is a bit confusing, since backends are mixed with core modules. For example, it's not obvious to know if `pandas/` is a backend, or core functionality. Many of the backends live in the root directory, and some are grouped, like in a directory `sql/`. This makes things even harder and difficult to maintain. Also, ibis uses attributes to access the backends (e.g. `ibis.spark.connect`), and `ibis.spark` in code seems to be the `ibis/spark/` module, but it's not, making things confusing, and even creating problems with pytest.

We're moving backends to a standardized structure, in the `backends/` directory, one backend at a time (each backend in a separate pull request). We started by the OmniSciDB backend, and its PR can be a reference for the changes that need to be made here: #2392

The exact tasks to perform are:
- Move the backend from its original location to `backends/` (if the backend is in a subdirectory, like `sql/`, ignore the subdirectory and just have the backend inside `backends/`, so `sql/sqlite/` -> `backends/sqlite`)
- Move the content of `backends/<your-backend>/api.py` to `backends/<your-backend>/__init__.py`. Backends are using `api.py` for historical reasons, but the public backend API should be in the standard `__init__.py` file of the module.
- Update `ibis/__init__.py` so the backend is imported from the new module (`from ibis.<your-backend> import api` -> `from ibis import <your-backend>`)
- Look for places in the code where your backend is being imported, and update the imports. For imports inside the same backend, please use relative imports (e.g. `import ibis.<your-backend>.whatever` -> `import .whatever`). See the reference PR for examples.
- Update the documentation (in `doc/source/backends/<your-backend>.rst`) so the API is generated from the right location.

After performing the changes, the CI should be green, except for the packaging build (which requires changes in the conda-forge feedstock). You can ignore that error, those changes will be performed later for all moved backends.

### PR DESCRIPTION
**Pull request by baogianghoangvu**
Closes #2420 

Please let me know if the PR needs any further additions. Thank you for reviewing!

### CODE REVIEW

File: ibis/__init__.py [hunk 1]
```diff
@@ -33,7 +33,7 @@
 
 with suppress(ImportError):
     # pip install ibis-framework[sqlite]
-    import ibis.sql.sqlite.api as sqlite  # noqa: F401
+    import ibis.backends.sqlite as sqlite  # noqa: F401
```
Comment (datapythonista) [inline ibis/__init__.py:36]:
```suggestion
    from ibis.backends import sqlite  # noqa: F401
```

File: ibis/backends/sqlite/__init__.py [hunk 2]
```diff
@@ -13,8 +13,8 @@
 # limitations under the License.
 
 
-from ibis.sql.sqlite.client import SQLiteClient
-from ibis.sql.sqlite.compiler import dialect, rewrites  # noqa: F401
+from ibis.backends.sqlite.client import SQLiteClient
```
Comment (datapythonista) [inline ibis/backends/sqlite/__init__.py:16]: Those are the imports that would be great to be relative:
```suggestion
from .client import SQLiteClient
```

Same for the others too.

File: ibis/backends/sqlite/tests/conftest.py [hunk 3]
```diff
@@ -2,7 +2,7 @@
 
 import pytest
 
-import ibis
+from ... import sqlite
```
Comment (datapythonista) [inline ibis/backends/sqlite/tests/conftest.py:5]: You can revert this. Sorry I was not clear, the imports that would be nice to have relative are the imports inside the backend that import things inside the backend.

File: ibis/backends/sqlite/tests/conftest.py [hunk 4]
```diff
@@ -17,7 +17,7 @@ def dbpath(data_directory):
 
 @pytest.fixture(scope='module')
 def con(dbpath):
-    return ibis.sqlite.connect(dbpath)
+    return sqlite.connect(dbpath)
```
Comment (datapythonista) [inline ibis/backends/sqlite/tests/conftest.py:20]: You won't need these changes, or the next ones if you revert the import.

File: ibis/backends/sqlite/tests/test_client.py [hunk 5]
```diff
@@ -10,6 +10,8 @@
 import ibis.expr.types as ir
 from ibis.util import guid
 
+from ... import sqlite
```
Comment (datapythonista) [inline ibis/backends/sqlite/tests/test_client.py:13]: same as before

**Review by datapythonista — COMMENTED** [review]
Thanks @baogianghoangvu, nice job, just few things that weren't clear regarding the imports, other than that looks good.

**Comment by baogianghoangvu** [comment]
Thank you so much @datapythonista for your detailed review. I've made the fixes accordingly. Sorry for my mix-up in the `tests`.

File: ibis/backends/sqlite/tests/conftest.py [hunk 6]
```diff
@@ -34,7 +34,7 @@ def dialect():
 
 @pytest.fixture
 def translate(dialect):
-    from ibis.sql.sqlite.compiler import SQLiteDialect
+    from ibis.backends.sqlite.compiler import SQLiteDialect
```
Comment (datapythonista) [inline ibis/backends/sqlite/tests/conftest.py:37]:
```suggestion
    from .sqlite.compiler import SQLiteDialect
```

**Review by datapythonista — COMMENTED** [review]
Thanks @baogianghoangvu, just one last import that I think it's better to be relative, but almost there.

**Comment by baogianghoangvu** [comment]
Thank you so much for your guidance. Hope I didn't take up too much of your time.

**Review by datapythonista — APPROVED** [review]
Thanks @baogianghoangvu, looks great. The CI failures are unrelated to your changes here, there is something broken in Ibis dependencies. I'll ask you to merge master once that is fixed, and we should be able to merge the PR afterwards. Thanks!

**Comment by baogianghoangvu** [comment]
I've rebased my branch to the latest ibis master, and updated the ci recipe for `sqlite`.

**Comment by datapythonista** [comment]
Great, feel free to ping me if we get a green build. This should be ready. Thanks!

**Comment by baogianghoangvu** [comment]
@datapythonista Hi! The previously failed check has succeeded. Though I'm not sure if the skipped and cancelled checks (impala-related) are expected or not.

**Comment by datapythonista** [comment]
Not sure why conda has problems in the impala build, it worked well in the last PRs that we merged. I restarted the CI for now, but I'll have a look at the problem (builds timeout after one hour, and conda is that slow in that build).

**Comment by datapythonista** [comment]
@baogianghoangvu looks like CI is read because of isort. Can you merge master into your branch again, and run `isort .` in the root please.

Let's see if we get a green build and we can finally merge this. Thanks!

**Comment by baogianghoangvu** [comment]
When I `isort .`, some files from directories other than `backends/sqlite` are changed as well. The changes seem harmless so I'll try pushing.

Thank you for your advice!

**Comment by baogianghoangvu** [comment]
Looks like my local isort doesn't conform to the required config. I'm reverting the isort commit to see if this caused the PySpark tests to fail.

**Comment by datapythonista** [comment]
You'll have to update isort. `conda env update` should do it.

**Comment by baogianghoangvu** [comment]
That was my mistake. I ran isort from the wrong environment. I hope it's good now.

**Comment by jreback** [comment]
thanks @baogianghoangvu

**Comment by baogianghoangvu** [comment]
Thank you @datapythonista for your guidance and review!

**Comment by datapythonista** [comment]
Thanks for your great work with this @baogianghoangvu. Feel free to work on other issues if you feel like it.

**Pull request merged**